### PR TITLE
PE-780: custom output folder name

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

File: src/commands/download_folder.ts [hunk 1]
```diff
@@ -26,10 +27,10 @@ new CLICommand({
...
 			const driveId = await ardrive.getDriveIdForFolderId(folderId);
 			const driveKey = await parameters.getDriveKey({ driveId: driveId });
-			await ardrive.downloadPrivateFolder({ folderId, destFolderPath, maxDepth, driveKey });
+			await ardrive.downloadPrivateFolder({ folderId, destFolderPath, customFolderName, maxDepth, driveKey });
 		} else {
 			const ardrive = cliArDriveAnonymousFactory({});
-			await ardrive.downloadPublicFolder({ folderId, destFolderPath, maxDepth });
+			await ardrive.downloadPublicFolder({ folderId, destFolderPath, customFolderName, maxDepth });
 		}
 		console.log(`Folder with ID "${folderId}" was successfully download to "${destFolderPath}"`);
```
Comment (matibat) [inline src/commands/download_folder.ts:35]: We are not logging the fill path here. 
```shell
$ yarn ardrive download-folder -f "47f5bde9-61ba-49c7-b409-1aa4a9e250f6" --max-depth 1 -l ./my-awesome-folder
Folder with ID "47f5bde9-61ba-49c7-b409-1aa4a9e250f6" was successfully download to "/home/matibat/ardriveapp/ardrive-cli"
```
Suggestion:
```suggestion
		const fullOutputPath = customFolderName ? `${destFolderPath}/${customFolderName}` : destFolderPath;
		console.log(`Folder with ID "${folderId}" was successfully download to "${fullOutputPath}"`);
```

File: src/utils/local_folder_path.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import { basename, dirname, resolve } from 'path';
+
+export type FilePathAndName = [string, string?]; // TODO: consolidate with the download-file functionality
+
+/**
+ * Resolves the path, verifies its existance and type to conditionally return its dir and file name
+ * @param {string} destOutputPath - the path from where to extract the dir path and name
+ * @returns {FilePathAndName} - the directory where to put the file and the file name (which is undefined when the provided destOutputPath is a directory)
+ */
+export function getOutputFolderPathAndName(destOutputPath: string): FilePathAndName {
```
Comment (matibat) [inline src/utils/local_folder_path.ts:11]: TODO: add unit tests for this method
Q: How do I stub `resolve`, `dirname`,`basename`, `statSync` and `Stats.isDirectory` here?

Comment (matibat) [reply src/utils/local_folder_path.ts:11]: Do I inject `fs` or an fs-like interface?

Comment (matibat) [reply src/utils/local_folder_path.ts]: The approach was to use injection as said: 6112bd8

File: src/commands/download_folder.ts [hunk 3]
```diff
@@ -5,18 +5,31 @@ import { CLIAction } from '../CLICommand/action';
...
 
 new CLICommand({
 	name: 'download-folder',
-	parameters: [FolderIdParameter, LocalPathParameter, MaxDepthParameter, ...DrivePrivacyParameters],
+	parameters: [
+		FolderIdParameter,
+		{
+			name: LocalFilePathDownloadParameter,
+			description:
+				'(OPTIONAL) the path on the local filesystem where the folder should be downloaded. Defaults to current working directory'
```
Comment (matibat) [inline src/commands/download_folder.ts:21]: Temporary overriding the description until we can multiple parameters witht the same name

File: src/utils/local_folder_path.test.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import { expect } from 'chai';
+import { join as joinPath } from 'path';
+import { stub } from 'sinon';
+import { getOutputFolderPathAndName } from './local_folder_path';
+
+const NAME_EXSTING_FOLDER_NAME = 'EXISTING_folder';
+const NAME_EXISTING_FILE = 'EXISTING_file.txt';
+// const NAME_EXISTING_NON_FILE = 'EXISTING_socket_or_symbolic_link';
```
Comment (matibat) [inline src/utils/local_folder_path.test.ts:8]: Remove the commented code

Comment (matibat) [reply src/utils/local_folder_path.test.ts]: 758613e96788d4ce6f0e07ec44aab0c97680d7eb

File: package.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
 	},
 	"types": "./lib/index.d.ts",
 	"dependencies": {
-		"ardrive-core-js": "../ardrive-core-js_PE-463_download_folder",
+		"ardrive-core-js": "../ardrive-core-js_PE-780_download_custom_folder_name",
```
Comment (arielmelendez) [inline package.json:11]: TODO: Update

File: src/commands/download_folder.ts (hunk 3 above)
Comment (arielmelendez) [reply src/commands/download_folder.ts:21]:
```suggestion
			description:
				'(OPTIONAL) the path on the local filesystem where the folder should be created and into which its contents are then downloaded. By default, the folder is created in the current working directory.'
```

File: src/commands/download_folder.ts [hunk 6]
```diff
@@ -5,18 +5,31 @@ import { CLIAction } from '../CLICommand/action';
 import {
 	DrivePrivacyParameters,
 	FolderIdParameter,
-	LocalPathParameter,
+	LocalFilePathDownloadParameter,
```
Comment (arielmelendez) [inline src/commands/download_folder.ts:8]: Consider refactoring to "DownloadLocalPathParameter"

Comment (matibat) [reply src/commands/download_folder.ts]: Applied

File: src/utils/local_path.ts [hunk 7]
```diff
@@ -45,3 +45,34 @@ export function getOutputFilePathAndName(
...
+	destOutputPath: string,
+	fsStatSyncAndPathResolveWrapper = fsStatSyncAndPathResolve
+): FilePathAndName {
+	const resolvedOutputPath = fsStatSyncAndPathResolveWrapper.resolve(destOutputPath);
+	const outputDirname = dirname(resolvedOutputPath);
+	const outputBasename = basename(resolvedOutputPath);
+	try {
+		const outputPathStats = fsStatSyncAndPathResolveWrapper.statSync(resolvedOutputPath);
+		// the destination does exist
+		if (outputPathStats.isDirectory()) {
```
Comment (matibat) [inline src/utils/local_path.ts:64]: Add a comment for the further PRs
```suggestion
		if (outputPathStats.isDirectory()) {
				// TODO: check case sensitivity conflicts here
```

**Review by matibat — COMMENTED** [review]
Left some comments regarding the logged error messages.

File: package.json (hunk 5 above)
Comment (matibat) [reply package.json]: The core version is currently set to an alpha version as in the PR base.

Closing this comment and adding at the base PR

File: src/utils/local_path.ts [hunk 8]
```diff
@@ -45,3 +45,34 @@ export function getOutputFilePathAndName(
 	}
 	throw new Error(`The destination is neither a folder nor a file!`);
 }
+
+/**
+ * Resolves the path, verifies its existance and type to conditionally return its dir and file name
+ * @param {string} destOutputPath - the path from where to extract the dir path and name
```
Comment (arielmelendez) [inline src/utils/local_path.ts:51]: params are out of date

**Review by arielmelendez — APPROVED** [review]
Expecting a few minor changes but should be good overall.

File: src/utils/local_path.ts (hunk 8 above)
Comment (matibat) [reply src/utils/local_path.ts:51]: a38b168c20b24f2c68c42faee952513d1f34a81e

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**